sklearn/partial_dependence.py

```diff
@@ -0,0 +1,577 @@
+                                     learning_rate, averaged_predictions[k])
+
+    return averaged_predictions
+
+
+def _partial_dependence_brute(est, grid, target_variables, X):
+
+    averaged_predictions = []
+    for new_values in grid:
+        X_eval = X.copy()
+        for i, variable in enumerate(target_variables):
+            X_eval[:, variable] = new_values[i]
+
+        try:
+            predictions = (est.predict(X_eval) if is_regressor(est)
+                           else est.predict_proba(X_eval))
+        except NotFittedError:
+            raise ValueError('est parameter must be a fitted estimator')
+
+        # Note: predictions is of shape
+        # (n_points,) for non-multioutput regressors
+        # (n_points, n_tasks) for multioutput regressors
+        # (n_points, 1) for the regressors in cross_decomposition (I think)
+        # (n_points, 2)  for binary classifaction
+        # (n_points, n_classes) for multiclass classification
+
+        # average over samples
+        averaged_predictions.append(np.mean(predictions, axis=0))
+
+    # reshape to (n_targets, n_points) where n_targets is:
+    # - 1 for non-multioutput regression and binary classification (shape is
+    #   already correct in those cases)
+    # - n_tasks for multi-output regression
+    # - n_classes for multiclass classification.
+    averaged_predictions = np.array(averaged_predictions).T
+    if is_regressor(est) and averaged_predictions.ndim == 1:
+        # non-multioutput regression, shape is (n_points,)
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+    elif is_classifier(est) and averaged_predictions.shape[0] == 2:
+        # Binary classification, shape is (2, n_points).
+        # we output the effect of **positive** class
+        averaged_predictions = averaged_predictions[1]
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+
+    return averaged_predictions
+
+
+def partial_dependence(est, target_variables, X, percentiles=(0.05, 0.95),
+                       grid_resolution=100, method='auto'):
+    """Partial dependence of ``target_variables``.
+
+    .. _warning_recursion_init:
+
+    .. warning::
+        The 'recursion' method only works for gradient boosting estimators,
+        and unlike the 'brute' method, it does not account for the ``init``
+        predictor of the boosting process. In practice this will produce the
+        same values as 'brute' up to a constant offset in the target
+        response, provided that ``init`` is a consant estimator (which is
+        the default). However, as soon as ``init`` is not a constant
+        estimator, the partial dependence values are incorrect.
+
+    Read more in the :ref:`User Guide <partial_dependence>`.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model. Multioutput-multiclass
+        classifiers are not supported.
+    target_variables : list or array-like of int
```

Comment:
can we just rename this to `features` or `columns` or `indices`?
Please state there must be at least two?

Reply:
Sure I'll rename to `features`.

It can be a list with a single element though, for classical one-way partial dependence. Do we want to also accepts ints?